src/lib.rs

```diff
@@ -584,15 +588,142 @@ pub trait Context<T, E>: context::private::Sealed {
         F: FnOnce() -> C;
 }
 
+static HOOK: OnceCell<ErrorHook> = OnceCell::new();
+
+#[cfg(feature = "std")]
+pub trait ReportHandler: core::any::Any + Send + Sync {
+    fn debug(
+        &self,
+        error: &(dyn StdError + 'static),
+        f: &mut core::fmt::Formatter<'_>,
+    ) -> core::fmt::Result;
+
+    #[cfg(backtrace)]
+    fn backtrace<'a>(&'a self, error: &'a (dyn StdError + 'static)) -> &'a Backtrace;
+
+    /// Override for the `Display` format
+    fn display(
+        &self,
+        error: &(dyn StdError + 'static),
+        f: &mut core::fmt::Formatter<'_>,
+    ) -> core::fmt::Result {
+        write!(f, "{}", error)?;
+
+        if f.alternate() {
+            for cause in crate::chain::Chain::new(error).skip(1) {
+                write!(f, ": {}", cause)?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(not(feature = "std"))]
+trait ReportHandler: core::any::Any + Send + Sync {
+    fn debug(
+        &self,
+        error: &(dyn StdError + 'static),
+        f: &mut core::fmt::Formatter<'_>,
+    ) -> core::fmt::Result;
+
+    #[cfg(backtrace)]
+    fn backtrace<'a>(&'a self, error: &'a (dyn StdError + 'static)) -> &'a Backtrace;
+
+    /// Override for the `Display` format
+    fn display(
+        &self,
+        error: &(dyn StdError + 'static),
+        f: &mut core::fmt::Formatter<'_>,
+    ) -> core::fmt::Result {
+        write!(f, "{}", error)?;
+
+        if f.alternate() {
+            for cause in crate::chain::Chain::new(error).skip(1) {
+                write!(f, ": {}", cause)?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+type ErrorHook = Box<
+    dyn Fn(&(dyn StdError + 'static)) -> Box<dyn ReportHandler + Send + Sync + 'static>
+        + Sync
+        + Send
+        + 'static,
+>;
+
+#[derive(Debug)]
+pub struct HookError;
+
+impl core::fmt::Display for HookError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.write_str("unable to set global hook")
+    }
+}
+
+#[cfg(feature = "std")]
```

Comment:
I think we might want to add some doc_cfg's for all these publicly exported `std` only apis.